File: frontend/src/lib/api/notifications.ts
/**
 * Notifications API Functions
 * Handles user notifications
 */

import api from "../api";

export interface Notification {
  id: string;
  user_id: string;
  type: "appointment" | "lab_result" | "prescription" | "system" | "message";
  title: string;
  message: string;
  read: boolean;
  action_url?: string;
  metadata?: Record<string, any>;
  created_at: string;
  read_at?: string;
}

export interface NotificationsListParams {
  unread_only?: boolean;
  type?: string;
  limit?: number;
  offset?: number;
}

export interface NotificationsListResponse {
  notifications: Notification[];
  total: number;
  unread_count: number;
}

/**
 * Get user notifications
 */
export async function getNotifications(
  params: NotificationsListParams = {}
): Promise<NotificationsListResponse> {
  const queryParams = new URLSearchParams();

  if (params.unread_only) {
    queryParams.append("unread_only", "true");
  }
  if (params.type) {
    queryParams.append("type", params.type);
  }
  if (params.limit) {
    queryParams.append("limit", params.limit.toString());
  }
  if (params.offset) {
    queryParams.append("offset", params.offset.toString());
  }

  const response = await api.get(`/api/v1/notifications?${queryParams}`);
  return response.data;
}

/**
 * Get unread notification count
 */
export async function getUnreadCount(): Promise<number> {
  const response = await api.get("/api/v1/notifications/unread-count");
  return response.data.count || 0;
}

/**
 * Mark notification as read
 */
export async function markAsRead(id: string): Promise<Notification> {
  const response = await api.post(`/api/v1/notifications/${id}/read`);
  return response.data;
}

/**
 * Mark notification as unread
 */
export async function markAsUnread(id: string): Promise<Notification> {
  const response = await api.post(`/api/v1/notifications/${id}/unread`);
  return response.data;
}

/**
 * Mark all notifications as read
 */
export async function markAllAsRead(): Promise<void> {
  await api.post("/api/v1/notifications/read-all");
}

/**
 * Delete notification
 */
export async function deleteNotification(id: string): Promise<void> {
  await api.delete(`/api/v1/notifications/${id}`);
}

/**
 * Delete all read notifications
 */
export async function deleteAllRead(): Promise<void> {
  await api.delete("/api/v1/notifications/read");
}

/**
 * Notification preference keys.
 * Both GET and PUT /notifications/preferences return the bare preferences map.
 * sms_notifications / whatsapp_notifications default to false; all others default to true.
 */
export interface NotificationPreferences {
  email_notifications: boolean;
  push_notifications: boolean;
  sms_notifications: boolean;
  whatsapp_notifications: boolean;
  appointment_reminders: boolean;
  queue_updates: boolean;
  lab_results: boolean;
  prescription_alerts: boolean;
  system_alerts: boolean;
}

export type NotificationPreferencesUpdate = Partial<NotificationPreferences>;

/**
 * Get notification preferences
 */
export async function getNotificationPreferences(): Promise<NotificationPreferences> {
  const response = await api.get("/api/v1/notifications/preferences");
  return response.data;
}

/**
 * Update notification preferences
 * Only the keys sent are updated; unset keys keep their current value.
 */
export async function updateNotificationPreferences(
  preferences: NotificationPreferencesUpdate
): Promise<NotificationPreferences> {
  const response = await api.put("/api/v1/notifications/preferences", preferences);
  return response.data;
}

/* ------------------------------------------------------------------------ */
/* Admin announcement broadcasts                                             */
/* ------------------------------------------------------------------------ */

export type BroadcastAudience = "all" | "patients" | "doctors" | "admins";
export type BroadcastType = "system" | "info" | "warning";

export interface BroadcastRequest {
  title: string;
  body: string;
  audience: BroadcastAudience;
  type?: BroadcastType;
  action_url?: string;
}

export interface BroadcastResult {
  sent: number;
  audience: BroadcastAudience;
  broadcast_id: string;
  title: string;
  target_role: "all" | "patient" | "doctor" | "admin";
  recipient_count: number;
  type: BroadcastType;
}

export interface BroadcastAudienceCount {
  audience: BroadcastAudience;
  count: number;
}

export interface BroadcastEntry {
  id: string;
  broadcast_id: string;
  title: string | null;
  body: string | null;
  audience: BroadcastAudience | null;
  target_role: "all" | "patient" | "doctor" | "admin" | null;
  type: BroadcastType | null;
  action_url: string | null;
  recipient_count: number | null;
  sent_by: string | null;
  sent_by_name: string | null;
  sent_at: string;
}

export interface BroadcastsResponse {
  data: BroadcastEntry[];
  total: number;
  page: number;
  limit: number;
  totalPages: number;
}

/**
 * Broadcast an announcement notification to every active user in the audience (admin only)
 */
export async function broadcastNotification(
  payload: BroadcastRequest
): Promise<BroadcastResult> {
  const response = await api.post(
    "/api/v1/admin/notifications/broadcast",
    payload
  );
  return response.data;
}

/**
 * Preview how many active users a broadcast audience targets (admin only)
 */
export async function getBroadcastAudienceCount(
  audience: BroadcastAudience
): Promise<BroadcastAudienceCount> {
  const response = await api.get(
    `/api/v1/admin/notifications/broadcast/count?audience=${audience}`
  );
  return response.data;
}

/**
 * List past announcement broadcasts, newest first (admin only)
 */
export async function listBroadcasts(
  page = 1,
  limit = 10
): Promise<BroadcastsResponse> {
  const response = await api.get(
    `/api/v1/admin/notifications/broadcasts?page=${page}&limit=${limit}`
  );
  return response.data;
}

File: backend/app/routers/notifications.py
from datetime import datetime, timezone
from uuid import UUID

from fastapi import APIRouter, Depends, HTTPException, Query, status
from sqlalchemy import and_, func, select, update
from sqlalchemy.ext.asyncio import AsyncSession

from app.database import get_db
from app.dependencies import get_current_user
from app.models.notification import Notification, NotificationPreferences, NotificationType
from app.models.user import User
from app.schemas.notification import (
    DeletedCountResponse,
    MarkAllReadResponse,
    NotificationPreferencesResponse,
    NotificationPreferencesUpdate,
    NotificationResponse,
    NotificationsListResponse,
    UnreadCountResponse,
)

router = APIRouter(prefix="/api/v1/notifications", tags=["notifications"])


DEFAULT_NOTIFICATION_PREFERENCES: dict = {
    "email_notifications": True,
    "push_notifications": True,
    "appointment_reminders": True,
    "lab_results": True,
    "prescription_alerts": True,
    "system_alerts": True,
    # Channel toggles consumed by services/notification_channels.py —
    # without these, sms/whatsapp are opt-out-impossible.
    "sms_notifications": False,
    "whatsapp_notifications": False,
}


# Endpoints
@router.get("", response_model=NotificationsListResponse)
async def get_notifications(
    unread_only: bool = Query(False, description="Filter to unread notifications only"),
    type: str | None = Query(None, description="Filter by notification type"),
    limit: int = Query(20, ge=1, le=100, description="Number of notifications to return"),
    offset: int = Query(0, ge=0, description="Number of notifications to skip"),
    user: User = Depends(get_current_user),
    db: AsyncSession = Depends(get_db),
):
    """Get user's notifications with optional filters."""
    # Build base query
    query = select(Notification).where(
        and_(
            Notification.user_id == user.id,
            Notification.deleted_at.is_(None),
        )
    )

    # Apply filters
    if unread_only:
        query = query.where(Notification.read.is_(False))

    if type:
        try:
            notification_type = NotificationType(type)
            query = query.where(Notification.type == notification_type)
        except ValueError:
            raise HTTPException(
                status_code=status.HTTP_400_BAD_REQUEST,
                detail={"error": {"code": "INVALID_TYPE", "message": f"Invalid notification type: {type}"}},
            )

    # Get total count
    count_query = select(func.count()).select_from(query.subquery())
    total_result = await db.execute(count_query)
    total = total_result.scalar_one()

    # Get unread count
    unread_query = select(func.count()).where(
        and_(
            Notification.user_id == user.id,
            Notification.deleted_at.is_(None),
            Notification.read.is_(False),
        )
    )
    unread_result = await db.execute(unread_query)
    unread_count = unread_result.scalar_one()

    # Apply pagination and ordering
    query = query.order_by(Notification.created_at.desc()).offset(offset).limit(limit)

    # Execute query
    result = await db.execute(query)
    notifications = result.scalars().all()

    return NotificationsListResponse(
        notifications=[NotificationResponse.from_orm(n) for n in notifications],
        total=total,
        unread_count=unread_count,
    )


@router.get("/unread-count", response_model=UnreadCountResponse)
async def get_unread_count(
    user: User = Depends(get_current_user),
    db: AsyncSession = Depends(get_db),
):
    """Get count of unread notifications."""
    query = select(func.count()).where(
        and_(
            Notification.user_id == user.id,
            Notification.deleted_at.is_(None),
            Notification.read.is_(False),
        )
    )
    result = await db.execute(query)
    count = result.scalar_one()

    return {"count": count}


@router.post("/{notification_id}/read", response_model=NotificationResponse)
async def mark_as_read(
    notification_id: UUID,
    user: User = Depends(get_current_user),
    db: AsyncSession = Depends(get_db),
):
    """Mark notification as read."""
    # Get notification (ensure it belongs to current user)
    query = select(Notification).where(
        and_(
            Notification.id == notification_id,
            Notification.user_id == user.id,
            Notification.deleted_at.is_(None),
        )
    )
    result = await db.execute(query)
    notification = result.scalar_one_or_none()

    if not notification:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail={"error": {"code": "NOT_FOUND", "message": "Notification not found"}},
        )

    # Mark as read
    notification.read = True
    notification.read_at = datetime.now(timezone.utc)
    await db.commit()
    await db.refresh(notification)

    return NotificationResponse.from_orm(notification)


@router.post("/{notification_id}/unread", response_model=NotificationResponse)
async def mark_as_unread(
    notification_id: UUID,
    user: User = Depends(get_current_user),
    db: AsyncSession = Depends(get_db),
):
    """Mark notification as unread."""
    # Get notification (ensure it belongs to current user)
    query = select(Notification).where(
        and_(
            Notification.id == notification_id,
            Notification.user_id == user.id,
            Notification.deleted_at.is_(None),
        )
    )
    result = await db.execute(query)
    notification = result.scalar_one_or_none()

    if not notification:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail={"error": {"code": "NOT_FOUND", "message": "Notification not found"}},
        )

    # Mark as unread
    notification.read = False
    notification.read_at = None
    await db.commit()
    await db.refresh(notification)

    return NotificationResponse.from_orm(notification)


@router.post("/read-all", response_model=MarkAllReadResponse)
async def mark_all_as_read(
    user: User = Depends(get_current_user),
    db: AsyncSession = Depends(get_db),
):
    """Mark all user's notifications as read."""
    stmt = (
        update(Notification)
        .where(
            and_(
                Notification.user_id == user.id,
                Notification.deleted_at.is_(None),
                Notification.read.is_(False),
            )
        )
        .values(read=True, read_at=datetime.now(timezone.utc))
    )
    await db.execute(stmt)
    await db.commit()

    return {"message": "All notifications marked as read"}


@router.delete("/read", response_model=DeletedCountResponse)
async def delete_all_read(
    user: User = Depends(get_current_user),
    db: AsyncSession = Depends(get_db),
):
    """Soft delete all read notifications."""
    stmt = (
        update(Notification)
        .where(
            and_(
                Notification.user_id == user.id,
                Notification.deleted_at.is_(None),
                Notification.read.is_(True),
            )
        )
        .values(deleted_at=datetime.now(timezone.utc))
    )
    result = await db.execute(stmt)
    await db.commit()

    deleted_count = result.rowcount

    return {"deleted_count": deleted_count}


@router.delete("/{notification_id}", status_code=status.HTTP_204_NO_CONTENT)
async def delete_notification(
    notification_id: UUID,
    user: User = Depends(get_current_user),
    db: AsyncSession = Depends(get_db),
):
    """Soft delete a notification."""
    query = select(Notification).where(
        and_(
            Notification.id == notification_id,
            Notification.user_id == user.id,
            Notification.deleted_at.is_(None),
        )
    )
    result = await db.execute(query)
    notification = result.scalar_one_or_none()

    if not notification:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail={"error": {"code": "NOT_FOUND", "message": "Notification not found"}},
        )

    notification.deleted_at = datetime.now(timezone.utc)
    await db.commit()

    return None


@router.get("/preferences", response_model=NotificationPreferencesResponse)
async def get_notification_preferences(
    user: User = Depends(get_current_user),
    db: AsyncSession = Depends(get_db),
):
    """Get user's notification preferences."""
    query = select(NotificationPreferences).where(
        NotificationPreferences.user_id == user.id,
    )
    result = await db.execute(query)
    preferences = result.scalar_one_or_none()

    if not preferences:
        # Return default preferences
        return DEFAULT_NOTIFICATION_PREFERENCES

    return preferences.preferences


@router.put("/preferences", response_model=NotificationPreferencesResponse)
async def update_notification_preferences(
    preferences: NotificationPreferencesUpdate,
    user: User = Depends(get_current_user),
    db: AsyncSession = Depends(get_db),
):
    """Create or update user's notification preferences."""
    # Check if preferences exist
    query = select(NotificationPreferences).where(
        NotificationPreferences.user_id == user.id,
    )
    result = await db.execute(query)
    existing_prefs = result.scalar_one_or_none()

    # Only the keys the client actually sent are updated; the rest are preserved.
    updates = preferences.model_dump(exclude_unset=True)

    if existing_prefs:
        # Update existing
        merged = {**(existing_prefs.preferences or {}), **updates}
        existing_prefs.preferences = merged
        existing_prefs.updated_at = datetime.now(timezone.utc)
        await db.commit()
        await db.refresh(existing_prefs)
        return existing_prefs.preferences
    else:
        # Create new
        new_prefs = NotificationPreferences(
            user_id=user.id,
            preferences={**DEFAULT_NOTIFICATION_PREFERENCES, **updates},
        )
        db.add(new_prefs)
        await db.commit()
        await db.refresh(new_prefs)
        return new_prefs.preferences
